File: client/api/preview.ts
import { triggerPromise } from "@trippler/tr_lib/client"
import { bridge } from "."
import { IsZoneFree } from ".."
import { settings, config } from "../../shared/constants"
import { trace } from "@trippler/tr_lib/shared"
import { createSingleVehicle, destroyCreatedVehicle } from "@trippler/tr_kit/client"

let isInPreview = false

const createPreviewCam = (key: unknown, netId: unknown) => {
  const vehicle = NetworkGetEntityFromNetworkId(netId)
  const cam = settings[key].VEHICLES.preview.cam
  bridge.drawText.show()
  isInPreview = true

  FreezeEntityPosition(PlayerPedId(), true)
  SetVehicleUndriveable(vehicle, true)

  DoScreenFadeOut(200)
  setTimeout(() => {
    DoScreenFadeIn(200)
  }, 500)
  camHandle = CreateCamWithParams("DEFAULT_SCRIPTED_CAMERA", cam.coords[0], cam.coords[1], cam.coords[2], cam.rotation.verticalrotate, cam.rotation.horizontalrotate, cam.rotation.left_n_right, cam.fov, false, 0)
  SetCamActive(camHandle, true)
  RenderScriptCams(true, true, 2000, true, true)
}

const clearPreviewCam = async (netId: unknown) => {
  bridge.drawText.hide()

  FreezeEntityPosition(PlayerPedId(), false)
  const success = await destroyCreatedVehicle(netId)
  if (!success) trace('failed to create the preview vehicle, reponse was: ' + success)

  DoScreenFadeOut(200)
  setTimeout(() => {
    DoScreenFadeIn(200)
  }, 500)
  RenderScriptCams(false, false, 2000, true, true)
  isInPreview = false
}

export const previewVehicle = async (key: unknown, index: unknown) => {
  const selectedConfig = settings[key].ITEM
  const configItems = config[selectedConfig]
  if (!isInPreview) {
    const isPreviewSessionBusy = await triggerPromise('sessionStatus')
    if (isPreviewSessionBusy) {
      bridge.notify('Preview service is currently busy, try again later', 'error')
      return
    }
    const isThePlaceClear = await IsZoneFree(settings[key].VEHICLES.preview.coords)
    if (!isThePlaceClear) {
      bridge.notify('Preview point is currently busy!', 'error')
      return
    }
    emitNet('lenix_vehicles:server:setPreviewSessionBusy', true)
    const [handle, netId] = await createSingleVehicle({
      hash: GetHashKey(configItems[index].vehicle),
      coords: settings[key].VEHICLES.preview.coords,
    })
    if (handle && netId) {
      createPreviewCam(key, netId)
      const tick = setTick(() => {
        if (IsControlJustReleased(0, 177)) {
          emitNet('lenix_vehicles:server:setPreviewSessionBusy', false)
          clearPreviewCam(netId);
          clearTick(tick);
        }
      });
    }
  }
}

File: lenixm/lenix_vehicles/client/api/preview.lua
local api <const> = require 'client/api/index'
local core <const> = require 'client/index'
local constants <const> = require 'shared/constants/index'
local kit <const> = require '@trippler/tr_kit/client'

local isInPreview = false
local camHandle = nil

local function createPreviewCam(key, netId)
	local vehicle = NetworkGetEntityFromNetworkId(netId)
	local cam = constants.settings[key].VEHICLES.preview.cam
	api.bridge.drawText.show()
	isInPreview = true

	FreezeEntityPosition(PlayerPedId(), true)
	SetVehicleUndriveable(vehicle, true)

	DoScreenFadeOut(200)
	SetTimeout(500, function()
		DoScreenFadeIn(200)
	end)

	camHandle = CreateCamWithParams("DEFAULT_SCRIPTED_CAMERA", cam.coords[1], cam.coords[2], cam.coords[3],
		cam.rotation.verticalrotate, cam.rotation.horizontalrotate, cam.rotation.left_n_right, cam.fov, false, 0)
	SetCamActive(camHandle, true)
	RenderScriptCams(true, true, 2000, true, true)
end

local function clearPreviewCam(netId)
	api.bridge.drawText.hide()
	FreezeEntityPosition(PlayerPedId(), false)

	local success = kit.destroyCreatedVehicle(netId)
	if not success then lib.trace('failed to destroy the preview vehicle') end

	DoScreenFadeOut(200)
	SetTimeout(500, function()
		DoScreenFadeIn(200)
	end)
	RenderScriptCams(false, false, 2000, true, true)
	isInPreview = false
end

local function previewVehicle(key, index)
	local selectedConfig = constants.settings[key].ITEM
	local configItems = constants.config[selectedConfig]

	if not isInPreview then
		local _, isPreviewSessionBusy = lib.triggerPromise('sessionStatus')
		if isPreviewSessionBusy then
			api.bridge.notify('Preview service is currently busy, try again later', 'error')
			return
		end

		local isThePlaceClear = core.IsZoneFree(constants.settings[key].VEHICLES.preview.coords)
		if not isThePlaceClear then
			api.bridge.notify('Preview point is currently busy!', 'error')
			return
		end

		TriggerServerEvent('lenix_vehicles:server:setPreviewSessionBusy', true)

		local handle, netId = kit.createSingleVehicle({
			hash = GetHashKey(configItems[index].vehicle),
			coords = constants.settings[key].VEHICLES.preview.coords,
		})

		if handle and netId then
			createPreviewCam(key, netId)
			-- Tick implementation using CreateThread
			CreateThread(function()
				while isInPreview do
					if IsControlJustPressed(0, 177) then -- Backspace/ESC
						clearPreviewCam(netId)
						TriggerServerEvent('lenix_vehicles:server:setPreviewSessionBusy', false)
						break
					end
					Wait(0)
				end
			end)
		end
	end
end

return {
	previewVehicle = previewVehicle
}
